File: products/chargeback/internal/rating/partners_test.go
package rating

import (
	"math/big"
	"testing"

	"github.com/openova-io/openova/products/chargeback/internal/store"
)

// The partner waterfall is decided by the ONE discount engine (DESIGN.md
// §Partners): a tier is a set of percent discounts off list, combined by the
// operator's combination rule exactly as a customer's discounts are. These
// tests pin that — no second pricing path, and the same rule table.

func tierPct(name, value, sku string) store.Discount {
	return store.Discount{ID: name, Name: name, Kind: "percent", Value: store.Decimal(value), SKU: sku, Active: true}
}

func unitLine(sku, amount string) store.RatedLine {
	return store.RatedLine{SKU: sku, Quantity: "1", Unit: "instance-hour", UnitPrice: store.Decimal(amount), Amount: store.Decimal(amount)}
}

// DiscountBySKU allocates the SAME total ApplyDiscounts computes, per meter —
// which is what makes a per-line buy and net possible without a second engine.
func TestDiscountBySKUAllocatesTheSameTotalPerMeter(t *testing.T) {
	lines := []store.RatedLine{unitLine("ecs.a", "50"), unitLine("evs.b", "50")}
	discounts := []store.Discount{tierPct("global", "10", ""), tierPct("on-a", "20", "ecs.a")}
	for _, tc := range []struct {
		rule        string
		total, a, b string
	}{
		// The DESIGN.md §2.11 table, per SKU: 10 % global + 20 % on A over
		// a 100 bill split 50/50.
		{store.DiscountRuleMostSpecific, "15", "10", "5"},
		{store.DiscountRuleHighest, "15", "10", "5"},
		{store.DiscountRuleStack, "20", "15", "5"},
		{store.DiscountRuleCompound, "19", "14", "5"},
	} {
		t.Run(tc.rule, func(t *testing.T) {
			total, perSKU, err := DiscountBySKU(lines, discounts, tc.rule)
			if err != nil {
				t.Fatal(err)
			}
			if got := roundRat(total, 6); got != dec(tc.total) {
				t.Fatalf("total = %s, want %s", got, tc.total)
			}
			if got := roundRat(perSKU["ecs.a"], 6); got != dec(tc.a) {
				t.Fatalf("ecs.a = %s, want %s", got, tc.a)
			}
			if got := roundRat(perSKU["evs.b"], 6); got != dec(tc.b) {
				t.Fatalf("evs.b = %s, want %s", got, tc.b)
			}
			// The parts always sum to the whole.
			sum := new(big.Rat).Add(perSKU["ecs.a"], perSKU["evs.b"])
			if sum.Cmp(total) != 0 {
				t.Fatalf("parts %s do not sum to the total %s", roundRat(sum, 6), roundRat(total, 6))
			}
			// And the total is exactly what ApplyDiscounts reports.
			applied, _, err := ApplyDiscounts(lines, discounts, tc.rule)
			if err != nil {
				t.Fatal(err)
			}
			if string(applied) != roundRat(total, 6) {
				t.Fatalf("ApplyDiscounts = %s, DiscountBySKU = %s", applied, roundRat(total, 6))
			}
		})
	}
}

func dec(s string) string {
	r, _ := new(big.Rat).SetString(s)
	return roundRat(r, 6)
}

// A tier is percent discounts off list, and the combination rule decides how
// two of them meet — the founder's rule, applied to the buy price.
func TestTierDiscountsObeyTheCombinationRule(t *testing.T) {
	list := []store.RatedLine{unitLine("ecs.a", "100")}
	tier := []store.Discount{tierPct("tier-base", "30", ""), tierPct("tier-on-a", "20", "ecs.a")}
	// most-specific: the SKU-scoped 20 % wins over the whole-bill 30 %, so
	// the buy is 80 — a narrower tier rule beats a broader one.
	total, _, err := DiscountBySKU(list, tier, store.DiscountRuleMostSpecific)
	if err != nil {
		t.Fatal(err)
	}
	if got := roundRat(total, 6); got != dec("20") {
		t.Fatalf("most-specific tier = %s, want 20", got)
	}
	// highest: 30 % wins wherever it is scoped.
	if total, _, err = DiscountBySKU(list, tier, store.DiscountRuleHighest); err != nil {
		t.Fatal(err)
	}
	if got := roundRat(total, 6); got != dec("30") {
		t.Fatalf("highest tier = %s, want 30", got)
	}
	// stack: 50 % off list.
	if total, _, err = DiscountBySKU(list, tier, store.DiscountRuleStack); err != nil {
		t.Fatal(err)
	}
	if got := roundRat(total, 6); got != dec("50") {
		t.Fatalf("stacked tier = %s, want 50", got)
	}
	// compound: 1 − 0.8 × 0.7 = 44 %.
	if total, _, err = DiscountBySKU(list, tier, store.DiscountRuleCompound); err != nil {
		t.Fatal(err)
	}
	if got := roundRat(total, 6); got != dec("44") {
		t.Fatalf("compound tier = %s, want 44", got)
	}
}

// The derived retail book: buy = list − tier, retail = base × (1 + markup),
// and a retail price below the buy price is REPORTED, never refused.
func TestDeriveItemsBaseBuyMarkupAndBelowBuyWarning(t *testing.T) {
	list := store.PriceBook{
		ID: "b1", Name: "list", Scope: store.LayerCloud, Currency: "OMR", AnnualDivisor: 8760, BillStopped: "compute",
		Items: []store.PriceItem{{SKU: "ecs.a", Unit: "instance-hour", UnitPrice: "100"}},
	}
	tier := []store.Discount{tierPct("tier", "30", "")}

	// base = buy (70) with a 5 % markup → 73.5, above the buy price.
	items, below, err := deriveItems(list, tier, store.RetailRule{Base: store.RetailBaseBuy, MarkupPct: "5"}, store.DiscountRuleMostSpecific)
	if err != nil {
		t.Fatal(err)
	}
	if len(items) != 1 || string(items[0].UnitPrice) != "73.50000000" {
		t.Fatalf("retail = %+v, want 73.50000000", items)
	}
	if len(below) != 0 {
		t.Fatalf("73.5 is above the buy price of 70, but it was reported below buy: %+v", below)
	}

	// base = list (100) with a 5 % markup → 105.
	items, _, err = deriveItems(list, tier, store.RetailRule{Base: store.RetailBaseList, MarkupPct: "5"}, store.DiscountRuleMostSpecific)
	if err != nil {
		t.Fatal(err)
	}
	if string(items[0].UnitPrice) != "105.00000000" {
		t.Fatalf("retail off list = %s, want 105.00000000", items[0].UnitPrice)
	}

	// A markup negative enough to price BELOW the buy price: base = buy,
	// −10 % → 63, under the 70 the partner pays us. Derived anyway, with
	// the warning: the rule is the partner's to set.
	items, below, err = deriveItems(list, tier, store.RetailRule{Base: store.RetailBaseBuy, MarkupPct: "-10"}, store.DiscountRuleMostSpecific)
	if err != nil {
		t.Fatal(err)
	}
	if string(items[0].UnitPrice) != "63.00000000" {
		t.Fatalf("retail = %s, want 63.00000000", items[0].UnitPrice)
	}
	if len(below) != 1 || below[0].SKU != "ecs.a" || string(below[0].Buy) != "70.00000000" || string(below[0].Retail) != "63.00000000" {
		t.Fatalf("below-buy = %+v, want ecs.a retail 63 under buy 70", below)
	}

	// An override is MOST SPECIFIC first: the SKU beats the service beats
	// the rule's default.
	rule := store.RetailRule{Base: store.RetailBaseBuy, MarkupPct: "5", Overrides: []store.RetailOverride{
		{Scope: "service", Key: "ecs", MarkupPct: "20"},
		{Scope: "sku", Key: "ecs.a", MarkupPct: "50"},
	}}
	items, _, err = deriveItems(list, tier, rule, store.DiscountRuleMostSpecific)
	if err != nil {
		t.Fatal(err)
	}
	if string(items[0].UnitPrice) != "105.00000000" { // 70 × 1.5
		t.Fatalf("sku override = %s, want 105.00000000", items[0].UnitPrice)
	}
	rule.Overrides = rule.Overrides[:1] // service only
	items, _, err = deriveItems(list, tier, rule, store.DiscountRuleMostSpecific)
	if err != nil {
		t.Fatal(err)
	}
	if string(items[0].UnitPrice) != "84.00000000" { // 70 × 1.2
		t.Fatalf("service override = %s, want 84.00000000", items[0].UnitPrice)
	}
}

func TestServiceOfSKU(t *testing.T) {
	for sku, want := range map[string]string{"ecs.s6.large.2": "ecs", "evs.ssd.gb": "evs", "plan.s": "plan", "eip": "eip", "k8s.vcpu": "k8s"} {
		if got := ServiceOf(sku); got != want {
			t.Errorf("ServiceOf(%q) = %q, want %q", sku, got, want)
		}
	}
}

// A retail rule is validated before it can derive anything.
func TestValidateRetailRule(t *testing.T) {
	if _, err := store.ValidateRetailRule(store.RetailRule{Base: "cost"}); err == nil {
		t.Fatal("an unknown base was accepted")
	}
	if _, err := store.ValidateRetailRule(store.RetailRule{Base: store.RetailBaseBuy, MarkupPct: "abc"}); err == nil {
		t.Fatal("a non-numeric markup was accepted")
	}
	if _, err := store.ValidateRetailRule(store.RetailRule{Base: store.RetailBaseList, MarkupPct: "-150"}); err == nil {
		t.Fatal("a markup below -100 % was accepted")
	}
	r, err := store.ValidateRetailRule(store.RetailRule{Base: store.RetailBaseList, Overrides: []store.RetailOverride{{Scope: "SKU", Key: " ecs.a ", MarkupPct: "10"}}})
	if err != nil {
		t.Fatal(err)
	}
	if string(r.MarkupPct) != "0" || r.Overrides[0].Scope != "sku" || r.Overrides[0].Key != "ecs.a" {
		t.Fatalf("normalised rule = %+v", r)
	}
	if _, err := store.ValidateRetailRule(store.RetailRule{Base: store.RetailBaseList, Overrides: []store.RetailOverride{
		{Scope: "sku", Key: "a", MarkupPct: "1"}, {Scope: "sku", Key: "a", MarkupPct: "2"},
	}}); err == nil {
		t.Fatal("a duplicate override was accepted")
	}
}

// allocate splits a SKU's figure across its lines so the parts sum EXACTLY to
// the whole — money never gains or loses a unit in the split.
func TestAllocateSplitsExactly(t *testing.T) {
	lines := []store.RatedLine{unitLine("a", "1"), unitLine("a", "1"), unitLine("a", "1")}
	parts := allocateBySKU(lines, map[string]*big.Rat{"a": big.NewRat(10, 1)}, func(l store.RatedLine) store.Decimal { return l.Amount })
	sum := new(big.Rat)
	for _, p := range parts {
		r, _ := new(big.Rat).SetString(string(p))
		sum.Add(sum, r)
	}
	if sum.Cmp(big.NewRat(10, 1)) != 0 {
		t.Fatalf("parts %v sum to %s, want 10", parts, sum.FloatString(6))
	}
}

File: products/chargeback/ui/src/lib/partners.test.ts
import { describe, expect, it } from 'vitest'
import type { MarginReport, PriceItem } from '../api/types'
import { belowBuyOf, marginByCustomer, marginPct, marginTable, marginTotals, markupFor, partnerTerms, previewRetail, serviceOf, validateRetailRule } from './partners'

// The console's derivations must equal the server's (DESIGN.md §11): a
// preview that disagrees with the book it saves is worse than no preview.

const items: PriceItem[] = [
  { sku: 'ecs.m7n.xlarge.8', unit: 'instance-hour', unit_price: 100 },
  { sku: 'evs.ssd.gb', unit: 'gb-hour', unit_price: 10 },
]

describe('the retail-rule preview', () => {
  it('base = buy with a 5 % markup: list 100 − tier 30 % = buy 70, retail 73.5', () => {
    const rows = previewRetail(items, 30, { base: 'buy', markup_pct: 5 })
    expect(rows[0].list).toBe(100)
    expect(rows[0].buy).toBe(70)
    expect(rows[0].retail).toBeCloseTo(73.5, 8)
    expect(rows[0].margin).toBeCloseTo(3.5, 8)
    expect(rows[0].belowBuy).toBe(false)
    // The second SKU follows the same rule at its own list price.
    expect(rows[1].buy).toBeCloseTo(7, 8)
    expect(rows[1].retail).toBeCloseTo(7.35, 8)
  })

  it('base = list with a 5 % markup is 105, whatever the partner pays', () => {
    const rows = previewRetail(items, 30, { base: 'list', markup_pct: 5 })
    expect(rows[0].retail).toBeCloseTo(105, 8)
    expect(rows[0].buy).toBe(70)
  })

  it('a markup negative enough prices BELOW the buy price, and says so', () => {
    const rows = previewRetail(items, 30, { base: 'buy', markup_pct: -10 })
    expect(rows[0].retail).toBeCloseTo(63, 8)
    expect(rows[0].belowBuy).toBe(true)
    expect(rows[0].margin).toBeCloseTo(-7, 8)
    const below = belowBuyOf(rows)
    expect(below).toHaveLength(2)
    expect(below[0]).toMatchObject({ sku: 'ecs.m7n.xlarge.8', buy_unit_price: 70 })
    expect(below[0].retail_unit_price).toBeCloseTo(63, 8)
  })

  it('never prices below zero', () => {
    expect(previewRetail(items, 0, { base: 'list', markup_pct: -100 })[0].retail).toBe(0)
  })

  it('an override is most specific first: sku beats service beats the default', () => {
    const rule = { base: 'buy' as const, markup_pct: 5, overrides: [{ scope: 'service', key: 'ecs', markup_pct: 20 }, { scope: 'sku', key: 'ecs.m7n.xlarge.8', markup_pct: 50 }] }
    expect(markupFor(rule, 'ecs.m7n.xlarge.8')).toBe(50)
    expect(markupFor(rule, 'ecs.other')).toBe(20)
    expect(markupFor(rule, 'evs.ssd.gb')).toBe(5)
    const rows = previewRetail(items, 30, rule)
    expect(rows[0].retail).toBeCloseTo(105, 8) // 70 × 1.5
    expect(rows[1].retail).toBeCloseTo(7.35, 8) // 7 × 1.05
  })

  it('names a service by the SKU it is the head of', () => {
    expect(serviceOf('ecs.s6.large.2')).toBe('ecs')
    expect(serviceOf('plan.s')).toBe('plan')
    expect(serviceOf('eip')).toBe('eip')
  })

  it('refuses a rule it cannot derive from', () => {
    expect(validateRetailRule({ base: 'buy', markup_pct: 5 })).toBe('')
    expect(validateRetailRule({ base: 'cost', markup_pct: 5 })).toMatch(/base/)
    expect(validateRetailRule({ base: 'buy', markup_pct: -150 })).toMatch(/−100/)
    expect(validateRetailRule({ base: 'buy', markup_pct: 5, overrides: [{ scope: 'sku', key: '', markup_pct: 1 }] })).toMatch(/override/)
    expect(
      validateRetailRule({ base: 'buy', markup_pct: 5, overrides: [{ scope: 'sku', key: 'a', markup_pct: 1 }, { scope: 'sku', key: 'a', markup_pct: 2 }] }),
    ).toMatch(/twice/)
  })
})

const report: MarginReport = {
  partner_id: 'p1',
  period: '2026-08',
  currency: 'OMR',
  rows: [
    { customer_id: 'c1', customer_name: 'Alpha', service: 'ecs', customer_net: 90, partner_buy: 70, margin: 20, margin_pct: 22.22 },
    { customer_id: 'c1', customer_name: 'Alpha', service: 'evs', customer_net: 10, partner_buy: 8, margin: 2, margin_pct: 20 },
    { customer_id: 'c2', customer_name: 'Bravo', service: 'ecs', customer_net: 50, partner_buy: 50, margin: 0, margin_pct: 0 },
  ],
  totals: { customer_id: '', customer_name: '', service: '', customer_net: 150, partner_buy: 128, margin: 22, margin_pct: 14.67 },
}

describe('the margin table', () => {
  it('derives the margin from net and buy, never from the sent figure', () => {
    const rows = marginTable(report)
    expect(rows).toHaveLength(3)
    expect(rows[0]).toMatchObject({ customer: 'Alpha', service: 'ecs', net: 90, buy: 70, margin: 20 })
    expect(rows[0].pct).toBeCloseTo(22.2222, 3)
    // A zero margin is a real answer, not a missing one.
    expect(rows[2].margin).toBe(0)
    expect(rows[2].pct).toBe(0)
  })

  it('totals what it shows', () => {
    const t = marginTotals(marginTable(report))
    expect(t.net).toBe(150)
    expect(t.buy).toBe(128)
    expect(t.margin).toBe(22)
    expect(t.pct).toBeCloseTo(14.6667, 3)
  })

  it('rolls up per customer, sorted by name', () => {
    const rows = marginByCustomer(marginTable(report))
    expect(rows.map((r) => r.customer)).toEqual(['Alpha', 'Bravo'])
    expect(rows[0]).toMatchObject({ net: 100, buy: 78, margin: 22 })
    expect(rows[0].pct).toBeCloseTo(22, 6)
  })

  it('has no percentage when the net is zero', () => {
    expect(marginPct(0, 0)).toBeNull()
    expect(marginTable({ ...report, rows: [{ customer_id: 'c', customer_name: 'C', service: 's', customer_net: 0, partner_buy: 0, margin: 0 }] })[0].pct).toBeNull()
  })

  it('reads an empty report as an empty table', () => {
    expect(marginTable(null)).toEqual([])
    expect(marginTotals([])).toMatchObject({ net: 0, buy: 0, margin: 0, pct: null })
  })
})

describe('the partner directory line', () => {
  it('says the model and what sets the buy price', () => {
    expect(partnerTerms({ id: 'p', slug: 'p', name: 'P', bill_to: 'partner', status: 'active', tier_name: 'Gold' })).toBe('Resell · Gold')
    expect(partnerTerms({ id: 'p', slug: 'p', name: 'P', bill_to: 'customer', status: 'active', commission_pct: 20 })).toBe('Agent · 20 % commission')
    expect(partnerTerms({ id: 'p', slug: 'p', name: 'P', bill_to: 'partner', status: 'active' })).toBe('Resell · no tier')
  })
})
